File: #23_Merge_K_Sorted_Lists/Solution.java
/* A solution based on min PQ */

import java.util.*;

class ListNode {
    int val;
    ListNode next;
    ListNode(int x) { val = x; }
}
 

public class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists == null || lists.length == 0) return null;
        ListNode dummyHead = new ListNode(0);
        ListNode cur = dummyHead;
        /* new comparator */
        PriorityQueue<ListNode> minPQ = new PriorityQueue<ListNode>(lists.length, 
        	new Comparator<ListNode>() {
        		@Override
        		public int compare(ListNode nd1, ListNode nd2) {
        			return nd1.val - nd2.val;
        		}
        });
        for (ListNode node : lists) {
        	if (node != null) {
        		minPQ.add(node);
        	}
        }
        if (minPQ.isEmpty()) return null;
        while (!minPQ.isEmpty()) {
        	ListNode min = minPQ.poll();
        	cur.next = min;
        	cur = min;
        	min = min.next;
        	if (min != null) minPQ.add(min);
        }
        return dummyHead.next;
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	ListNode head1 = new ListNode(1);
    	ListNode nd1= new ListNode(5);
    	ListNode nd2 = new ListNode(10);
    	head1.next = nd1;
    	nd1.next = nd2;
    	ListNode head2 = new ListNode(3);
    	ListNode nd3= new ListNode(7);
    	ListNode nd4 = new ListNode(11);
    	head2.next = nd3;
    	nd3.next = nd4;
    	ListNode[] lists = new ListNode[2];
    	lists[1] = head1;
    	lists[0] = head2;
    	ListNode result = sol.mergeKLists(lists);
    	while (true) {
    		System.out.println(result.val);
    		result = result.next;
    		if (result == null) break;
    	}
    }
}

File: #23_Merge_K_Sorted_Lists/Solution.py
'''
Priority queue based Solution

time: O(log(k) * n), space O(k)
'''

# import Queue as Q

class ListNode:
	def __init__(self, x):
		self.val = x
		self.next = None

# class Solution:
# 	# @param {ListNode[]} lists
# 	# @return {ListNode}
# 	def mergeKLists(self, lists):
# 		if not lists:
# 			return None
# 		if len(lists) == 1:
# 			return lists[0]
# 		dummy = ListNode(-1)
# 		ptr = dummy
# 		pq = Q.PriorityQueue()
# 		for head in lists:
# 			if head:
# 				pq.put((head.val, head))
# 		while not pq.empty():
# 			head = pq.get()[1]
# 			ptr.next = head
# 			head = head.next
# 			if head:
# 				pq.put((head.val, head))
# 			ptr = ptr.next
# 		return dummy.next

'''
Sort all nodes
time: O(log(n) * n), space(n)
This is practically faster because of partially sorted sequence
'''

# class Solution:
# 	# @param {ListNode[]} lists
# 	# @return {ListNode}
# 	def mergeKLists(self, lists):
# 		if not lists:
# 			return None
# 		if len(lists) == 1:
# 			return lists[0]
# 		sortedList = []
# 		for head in lists:
# 			while head:
# 				sortedList.append(head)
# 				head = head.next
# 		sortedList.sort(key=lambda ListNode: ListNode.val)
# 		for i in range(len(sortedList) - 1):
# 			sortedList[i].next = sortedList[i + 1]
# 		if not sortedList:
# 			return None
# 		return sortedList[0]

'''
divide and conquer
time: O(log(k) * n), space O(1)
'''

class Solution:
	# @param {ListNode[]} lists
	# @return {ListNode}
	def mergeKLists(self, lists):
		if not lists:
			return None
		if len(lists) == 1:
			return lists[0]
		mid = len(lists) / 2
		return self.merge2List(self.mergeKLists(lists[:mid]), self.mergeKLists(lists[mid:]))

	def merge2List(self, l1, l2):
		if not l1:
			return l2
		if not l2:
			return l1
		dummy = ListNode(-1)
		cur = dummy
		while l1 or l2:
			if not l1:
				node = l2
			elif not l2:
				node = l1
			else:
				if l1.val < l2.val:
					node = l1
				else:
					node = l2
			cur.next = node
			if node == l1:
				l1 = l1.next
			else:
				l2 = l2.next
			cur = cur.next
		return dummy.next

sol = Solution()
sol.mergeKLists([None])